src/blog/articles/javascript-async-iterators/index.md

```diff
@@ -0,0 +1,689 @@
+---
+date: 2021-05-04T18:00:00
+updatedAt: 2021-05-04T18:00:00
+title: JavaScript async iterators
+permalink: /blog/javascript-async-iterators/
+description: An in-depth exploration of JavaScript iteration protocols with a special focus on async iterators
+layout: article
+author: Luciano Mammino
+author_profile_pic: /static/luciano-mammino-avatar-small.jpg
+author_link: https://loige.co/
+tags: ["blog"]
+---
+
+Did you know that JavaScript offers a few protocols to allow iteration over certain objects? Of course, we know we can easily iterate over arrays, but with these protocols, you can make your own custom objects iterable as well.
+
+When you have an iterable object representing a collection, you can use the `for...of` syntax to iterate over every single item of the collection.
+
+But what if an object abstracts data generated asynchronously? For instance, think of an abstraction that allows us to fetch data from a paginated API, or records consumed in batches from a database, or something as simple as a countdown timer. Well in these cases you can use the `for await...of` syntax!
+
+In this article, we will learn more about the _iterator_ and the _iterable_ protocol (and their async counterparts) and we will see how to create custom objects that can expose their internal data in an ergonomic and idiomatic way.
+
+
+## JavaScript iteration with `for...of`
+
+With ECMAScript 2015, JavaScript got the `for...of` syntax. This syntax provides a very easy way to iterate over collections, like arrays, string, sets, and maps.
+
+If you have never seen this syntax in action here are some examples:
+
+```javascript
+const judokas = [
+  'Driulis Gonzalez Morales',
+  'Ilias Iliadis',
+  'Tadahiro Nomura',
+  'Anton Geesink',
+  'Teddy Riner',
+  'Ryoko Tani'
+]
+
+for (const judoka of judokas) {
+  console.log(judoka)
+}
+```
+
+In the example above, we are iterating over an array using the `for...of` syntax. If we run this code, this is what we will get as output:
+
+```text
+Driulis Gonzalez Morales
+Ilias Iliadis
+Tadahiro Nomura
+Anton Geesink
+Teddy Riner
+Ryoko Tani
+```
+
+The same syntax works also for iterating over the characters of a string:
+
+```javascript
+const judoka = 'Ryoko Tani'
+
+for (const char of judoka) {
+  console.log(char)
+}
+```
+
+This will print:
```

Comment:
> **The above** will print

There are multiple instances of "This will output" or similar. I'd make it more clear with something like "The above" or the "previous code". "This" may even refer to the next output or code.